tgp-2prpl.c

```diff
@@ -193,6 +197,88 @@ int p2tgl_imgstore_add_with_id_raw (const unsigned char *raw_bgra, unsigned widt
   return purple_imgstore_add_with_id (tga, tga_len, NULL);
 }
 
+#ifdef HAVE_LIBPNG
+
+struct mem_encode
+{
+  char *buffer;
+  size_t size;
+};
+
+static void p2tgl_png_mem_write (png_structp png_ptr, png_bytep data, png_size_t length) {
+  struct mem_encode *p = (struct mem_encode *) png_get_io_ptr(png_ptr);
+  size_t nsize = p->size + length;
+  
+  if (p->buffer)
+    p->buffer = g_realloc(p->buffer, nsize);
+  else
+    p->buffer = g_malloc(nsize);
+  
+  memcpy(p->buffer + p->size, data, length);
+  p->size = nsize;
+}
+
+int p2tgl_imgstore_add_with_id_png (const unsigned char *raw_bitmap, unsigned width, unsigned height) {
+  struct mem_encode state = {NULL, 0};
+  png_structp png_ptr = NULL;
+  png_infop info_ptr = NULL;
+  png_bytepp rows = NULL;
+
+  // init png write struct
+  png_ptr = png_create_write_struct (PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  if (png_ptr == NULL) {
+    warning ("error encoding png (create_write_struct failed)");
+    return 0;
+  }
+  
+  // init png info struct 
+  info_ptr = png_create_info_struct (png_ptr);
+  if (info_ptr == NULL) {
+    png_destroy_write_struct(&png_ptr, NULL);
+    warning ("error encoding png (create_info_struct failed)");
+    return 0;
+  }
+  
+  // Set up error handling.
+  if (setjmp(png_jmpbuf(png_ptr))) {
+    png_destroy_write_struct(&png_ptr, &info_ptr);
+    warning ("error while writing png");
+    return 0;
+  }
+  
+  // set img attributes
+  png_set_IHDR (png_ptr, info_ptr, width, height, 
+                8, PNG_COLOR_TYPE_RGBA, PNG_INTERLACE_NONE,
+                PNG_COMPRESSION_TYPE_DEFAULT, PNG_FILTER_TYPE_DEFAULT);
+  
+  // alloc row pointers
+  rows = (png_bytepp) malloc(height * sizeof(png_bytep));
```

Comment:
`rows` is not managed here by libpng, so it's possible to use any mm, I've just used most generic.